src/server/pfs/server/driver.go

```diff
@@ -2451,71 +2454,109 @@ func (d *driver) getFile(pachClient *client.APIClient, file *pfs.File, offset in
+				if err != nil {
+					return fmt.Errorf("file %q has a header, but could not "+
+						"retrieve parent node at %q to get header content: %v", p,
+						parentPath, err)
+				}
+				if parentNode.DirNode == nil {
+					return fmt.Errorf("parent of %q is not a directory—this is "+
+						"likely an internal error", p)
+				}
+				if parentNode.DirNode.Shared == nil {
+					return fmt.Errorf("file %q has a shared header or footer, "+
+						"but parent directory does not permit shared data", p)
+				}
+				if parentNode.DirNode.Shared.Header != nil {
+					objects = append(objects, parentNode.DirNode.Shared.Header)
+				}
+				if parentNode.DirNode.Shared.Footer != nil {
+					footer = parentNode.DirNode.Shared.Footer
+				}
+			}
+		}
+		return collectReferences(p, node)
+	}
+
 	// Handle commits to input repos
 	if !provenantOnInput(commitInfo.Provenance) {
 		tree, err := d.getTreeForFile(pachClient, client.NewFile(file.Commit.Repo.Name, file.Commit.ID, ""))
 		if err != nil {
 			return nil, err
 		}
-		var (
-			pathsFound int
-			objects    []*pfs.Object
-			totalSize  uint64
-			footer     *pfs.Object
-			prevDir    string
-		)
 		if err := tree.Glob(file.Path, func(p string, node *hashtree.NodeProto) error {
 			pathsFound++
-			if node.FileNode == nil {
-				return nil
-			}
-
-			// add footer + header for next dir. If a user calls e.g.
-			// 'GetFile("/*/*")', then the output looks like:
-			// [d1 header][d1/1]...[d1/n][d1 footer] [d2 header]...[d2 footer] ...
-			parentPath := path.Dir(p)
-			if parentPath != prevDir {
-				if footer != nil {
-					objects = append(objects, footer)
-				}
-				footer = nil // don't apply footer twice if next dir has no footer
-				prevDir = parentPath
-				if node.FileNode.HasHeaderFooter {
-					// if any child of 'node's parent directory has HasHeaderFooter set,
-					// then they all should
-					parentNode, err := tree.Get(parentPath)
-					if err != nil {
-						return fmt.Errorf("file %q has a header, but could not "+
-							"retrieve parent node at %q to get header content: %v", p,
-							parentPath, err)
-					}
-					if parentNode.DirNode == nil {
-						return fmt.Errorf("parent of %q is not a directory—this is "+
-							"likely an internal error", p)
-					}
-					if parentNode.DirNode.Shared == nil {
-						return fmt.Errorf("file %q has a shared header or footer, "+
-							"but parent directory does not permit shared data", p)
-					}
-					if parentNode.DirNode.Shared.Header != nil {
-						objects = append(objects, parentNode.DirNode.Shared.Header)
-					}
-					if parentNode.DirNode.Shared.Footer != nil {
-						footer = parentNode.DirNode.Shared.Footer
-					}
+			if node.DirNode != nil {
+				if recursive {
+					return tree.Walk(node.GetName(), attachHeaderFooterToMetadata)
 				}
+				return nil
 			}
-			objects = append(objects, node.FileNode.Objects...)
-			totalSize += uint64(node.SubtreeSize)
-			return nil
+			return attachHeaderFooterToMetadata(p, node)
```

Comment:
I think the call chain "attachHeaderFooterToMetadata -> collectReferences" is a little misleading in that you have to go through attachHeaderFooterToMetadata even if there is no header or footer. It might be clearer to have a check for a header/footer that will be used to conditionally call a function that attaches the header/footer, then call collect references after the if block.

Reply:
I definitely agree that it's misleading to call `collectReferences` from within `attachHeaderFooter`, I have pulled that out to have it called separately. 

I'm not sure we want to conditionally call an attachHeaderFooter function, because we are then littering the getFile function with multiple copies (both inside and outside the walk callback) of tree getting, parent-path, and previous footer logic that are really only used for the dealing with headers and footers. Also there are a few levels of conditionals (whether the parent path matches, whether the footer is nil or not from the last iteration, whether the current node has header/footer), and picking any one of them to be the one to trigger the call to separate function is kind of arbitrary so for the sake of duplicating as little code in the main body of `getFile` it seems cleanest to isolate all the header/footer logic into one single function. Happy to break it out if you disagree, but it seemed like the main confusion comes from going through the header/footer code to call `collectReferences`.